In [1]:
! rm -r /content/SearchQuery2FuncCall
!git clone https://github.com/XiaoLIUau/SearchQuery2FuncCall.git

rm: cannot remove '/content/SearchQuery2FuncCall': No such file or directory
Cloning into 'SearchQuery2FuncCall'...
remote: Enumerating objects: 106, done.
remote: Counting objects: 100% (106/106), done.
remote: Compressing objects: 100% (71/71), done.
remote: Total 106 (delta 67), reused 71 (delta 34), pack-reused 0
Receiving objects: 100% (106/106), 78.48 KiB | 1.48 MiB/s, done.
Resolving deltas: 100% (67/67), done.


In [2]:
%pip install --upgrade pip
%pip install --disable-pip-version-check \
    torch==1.13.1 \
    torchdata==0.5.1

%pip install --upgrade accelerate\
             --upgrade huggingface_hub

%pip install \
    bitsandbytes>0.37.2 \
    transformers==4.28.1 \
    datasets==2.11.0 \
    evaluate==0.4.0 \
    rouge_score==0.1.2 \
    loralib==0.1.1 \
    peft==0.3.0 \
    trl==0.4.7 --quiet


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.1/2.1 MB 10.5 MB/s eta 0:00:00
  Attempting uninstall: pip
    Found existing installation: pip 23.1.2
    Uninstalling pip-23.1.2:
      Successfully uninstalled pip-23.1.2
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 887.5/887.5 MB 1.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.6/4.6 MB 105.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 849.3/849.3 kB 70.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 557.1/557.1 MB 3.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 317.1/317.1 MB 4.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.0/21.0 MB 80.9 MB/s eta 0:00:00
  Obtaining dependency information for portalocker>=2.0.0 from https://files.pythonhosted.org/packages/17/9e/87671efcca80ba6203811540ed1f9c0462c1609d2281d7b7f53cef05da3d/portalocker-2.8.2-py3-none-any.whl.metadata
  Attempting uninstall: torch
    Found existing installation: torch 2.0.1

# Setup

In [ ]:
# load libraries
import os
import time
import pandas as pd
import torch
import evaluate
import tqdm

from transformers import AutoModelForCausalLM, AutoTokenizer, BitsAndBytesConfig, GenerationConfig, TrainingArguments, Trainer
from peft import PeftModel, LoraConfig, get_peft_model, TaskType
from SearchQuery2FuncCall.setup_dataset import text2json, load_n_process_data


### Load datasets
text2json('/content/SearchQuery2FuncCall/Dataset.txt')
# q2f_datasets = load_n_process_data('/content/non_search_examples.json')
q2f_datasets = load_n_process_data('/content/q2f_dataset.json')
q2f_datasets


### Access Huggingface for loading model
""" # Get model api key """
def load_api_key_from_file(file_path):
  with open(file_path, 'r') as file:
      api_key = file.read().strip()
  return api_key


# Setting a new environment variable
os.environ["HUGGINGFACE_TOKEN"] = load_api_key_from_file('/content/api_key_huggingface.txt')
!huggingface-cli login --token $HUGGINGFACE_TOKEN


### Funciton to print number of trainable mode parameters
def print_number_of_trainable_model_parameters(model):
  trainable_model_params = 0
  all_model_params = 0
  for _, param in model.named_parameters():
      all_model_params += param.numel()
      if param.requires_grad:
          trainable_model_params += param.numel()
  return f"trainable model parameters: {trainable_model_params}\nall model parameters: {all_model_params}\npercentage of trainable model parameters: {100 * trainable_model_params / all_model_params:.2f}%"


### Prepare for model generation
# define prompt format
def create_prompt(input):
  input_prompt = 'Input:'
  output_prompt = ', Output:'
  instruction = f"""Instruction: Given a search query, then route to different backend components based on the search intent.
1. If the search is about unit conversion, return API function UnitConvert(SourceUnit, TargetUnit, SourceValue).
2. If the search is about calculation, return API function Calculate(Equation).
3. If the search is about other search intent, return API function Search().
* For unit conversion: common unit conversion in length, mass, time, area, speed, temperature, volume should be covered. And it should be consistent for the same unit throughout. E.g. it should always be “foot”, it cannot be “feet” or “ft” in API calls.
* For calculation: common operation such as +, -, *, /, pow, log, ln, exp, tan(h), sin(h), cos(h), factorial should be covered. And it should be consistent for the same operation throughout. E.g. it should always be “ * ”, it cannot be “x” or “X” in API calls.
Handle input queries in different language styles. Cover common unit conversion and calculation operations.

"""
  prompt = instruction + input_prompt + f'“{input}”' + output_prompt
  return prompt

# Generate response in tokens with given model and tokenized input
def model_generate(model, input):
  generation_config = model.generation_config
  generation_config.max_new_tokens = 100
  generation_config.temperature = 0.00000000000001
  generation_config.top_p = 0.9
  generation_config.num_return_sequences = 1
  generation_config.pad_token_id = tokenizer.eos_token_id
  generation_config.eos_token_id = tokenizer.eos_token_id
  return model.generate(
          input_ids = input.input_ids,
          attention_mask = input.attention_mask,
          generation_config = generation_config,
          )

# Generate Text with selected model and input text
def generated_text(model,input):
  prompt = create_prompt(input)
  inputs = tokenizer(prompt, return_tensors='pt')
  generated = tokenizer.decode(
      model_generate(model, inputs)[0],
      skip_special_tokens=True
  )
  return generated

# Post generation string processing
def extractOutputString(input_string,output_string):
  import re
  # Use regular expressions to find the matching output for the input query
  # output_match = re.search(rf'Input:\s*“{re.escape(input_string)}”\s*,\s*Output:\s*(.*)\[\/', output_string, flags=re.MULTILINE)
  output_match = re.search(rf'Input:\s*“{re.escape(input_string)}”\s*,\s*Output:\s*(.*?)(\[\/|$)', output_string, flags=re.MULTILINE)

  # Extract and print the output
  if output_match:
      output_string = output_match.group(1)
  # else:
  #     output_match = re.search(rf'Input:\s*“{re.escape(input_string)}”\s*,\s*Output:\s*(.*)$', output_string, flags=re.MULTILINE)
  #     if output_match:
  #         output_string = output_match.group(1)

  # Remove quotation marks
  prefixes = ['“', '”', "'", '"', '[', '.', ']']
  if output_string.startswith(tuple(prefixes)):
      output_string = output_string[1:]
  while output_string.endswith(tuple(prefixes)):
      output_string = output_string[:-1]
  # Remove all space in output
  output_string = "".join(output_string.split())
  return output_string



# Setup for test model output and dataset evaluation using ROUGE and BLEU scores

In [ ]:
def generate_samples(example_indices,datasets,model):
    dash_line = '-'.join('' for x in range(100))

    for i, index in enumerate(example_indices):
        input = datasets['test'][index]['input']
        output = datasets['test'][index]['output']

        generated = generated_text(model,input)
        generated = extractOutputString(input,generated)

        print(dash_line)
        print('Example ', i + 1)
        print(dash_line)
        print(f'INPUT:\n{input}')
        print(dash_line)
        print(f'BASELINE OUTPUT:\n{output}')
        print(dash_line)
        print(f'PEFT MODEL GENERATION - OUTPUT:\n{generated}\n')
    return


def generate_dataset(start_index,datasets,model):
    end_index=start_index+len(datasets['test'])
    inputs = datasets['test'][start_index:end_index]['input']
    outputs = datasets['test'][start_index:end_index]['output']

    outputs_gen = []

    for idx, input in enumerate(inputs):
        if idx%10==0:
            print(idx)
        output_gen = generated_text(model,input)
        output_gen = extractOutputString(input,output_gen)
        outputs_gen.append(output_gen)


    zipped_summaries = list(zip(inputs, outputs, outputs_gen))
    df = pd.DataFrame(zipped_summaries, columns = ['Inputs', 'Outputs', 'Outputs_generated'])
    return df


def evaluate_generations(outputs_ref,outputs_gen):
    # Rouge
    rouge = evaluate.load('rouge')
    model_results = rouge.compute(
        predictions=outputs_gen,
        references=outputs_ref,
        use_aggregator=True,
        use_stemmer=True,
    )

    print('MODEL ROUGE SCORES:')
    print(model_results)

    # bleu
    bleu = evaluate.load('bleu')
    model_results = bleu.compute(
        predictions=outputs_gen,
        references=outputs_ref,
    )

    print('MODEL BLEU SCORES:')
    print(model_results)

# Setup for training

In [3]:
import os
import time
import pandas as pd
import torch
import evaluate
import tqdm

from transformers import AutoModelForCausalLM, AutoTokenizer, BitsAndBytesConfig, GenerationConfig, TrainingArguments, Trainer

from peft import PeftModel, LoraConfig, get_peft_model, TaskType

# Load and process dataset

In [4]:
from SearchQuery2FuncCall.setup_dataset import text2json, load_n_process_data

text2json('/content/SearchQuery2FuncCall/Dataset.txt')
# q2f_datasets = load_n_process_data('/content/non_search_examples.json')
q2f_datasets = load_n_process_data('/content/q2f_dataset.json')
q2f_datasets

Saved 340 examples to 'q2f_dataset.json'.
Separated 87 Search() examples to 'search_examples.json'.
Separated 253 non-Search() examples to 'non_search_examples.json'.


Extracting data files:   0%|          | 0/1 [00:00<?, ?it/s]

Generating train split: 0 examples [00:00, ? examples/s]

Dataset json downloaded and prepared to /root/.cache/huggingface/datasets/json/default-60a145534b3c72b3/0.0.0/fe5dd6ea2639a6df622901539cb550cf8797e5a6b2dd7af1cf934bed8e233e6e. Subsequent calls will reuse this data.


  0%|          | 0/1 [00:00<?, ?it/s]

DatasetDict({
    train: Dataset({
        features: ['input', 'output'],
        num_rows: 210
    })
    validation: Dataset({
        features: ['input', 'output'],
        num_rows: 50
    })
    test: Dataset({
        features: ['input', 'output'],
        num_rows: 80
    })
})

# Load Huggingface API key
Here we use Huggingface models

Note: Please load a text file that contains your model api key to current folder

>Name your file in either ***'api_key_huggingface.txt'***


In [5]:
""" # Get model api key """
def load_api_key_from_file(file_path):
    with open(file_path, 'r') as file:
        api_key = file.read().strip()
    return api_key

# Setting a new environment variable
os.environ["HUGGINGFACE_TOKEN"] = load_api_key_from_file('/content/api_key_huggingface.txt')

!huggingface-cli login --token $HUGGINGFACE_TOKEN

Token will not been saved to git credential helper. Pass `add_to_git_credential=True` if you want to set the git credential as well.
Token is valid (permission: read).
Your token has been saved to /root/.cache/huggingface/token
Login successful


# Load Model and Setup Configs

In [6]:
# Free GPU memory
torch.cuda.empty_cache()

# Load model function
def load_model(model_name):
    bnb_config = BitsAndBytesConfig(
        load_in_8bit=True,
        bnb_8bit_use_double_quant=True,
        bnb_8bit_quant_type="nf4",
        bnb_8bit_compute_dtype=torch.bfloat16,
        llm_int8_enable_fp32_cpu_offload=True
    )
    return AutoModelForCausalLM.from_pretrained(
                    model_name,
                    device_map='auto',
                    quantization_config=bnb_config,
                    )

# Load model directly
model_name = "atwine/llama-2-7b-chat-fully-quantized-q4-06092023"
tokenizer = AutoTokenizer.from_pretrained(model_name)
tokenizer.pad_token = tokenizer.eos_token
tokenizer.truncation_side = "left"

# Original model
original_model = load_model(model_name)

(…)92023/resolve/main/tokenizer_config.json:   0%|          | 0.00/762 [00:00<?, ?B/s]

tokenizer.model:   0%|          | 0.00/500k [00:00<?, ?B/s]

(…)-q4-06092023/resolve/main/tokenizer.json:   0%|          | 0.00/1.84M [00:00<?, ?B/s]

(…)023/resolve/main/special_tokens_map.json:   0%|          | 0.00/414 [00:00<?, ?B/s]

(…)zed-q4-06092023/resolve/main/config.json:   0%|          | 0.00/991 [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/7.01G [00:00<?, ?B/s]

Some weights of LlamaForCausalLM were not initialized from the model checkpoint at atwine/llama-2-7b-chat-fully-quantized-q4-06092023 and are newly initialized: ['model.layers.10.self_attn.rotary_emb.inv_freq', 'model.layers.13.self_attn.rotary_emb.inv_freq', 'model.layers.26.self_attn.rotary_emb.inv_freq', 'model.layers.27.self_attn.rotary_emb.inv_freq', 'model.layers.6.self_attn.rotary_emb.inv_freq', 'model.layers.28.self_attn.rotary_emb.inv_freq', 'model.layers.29.self_attn.rotary_emb.inv_freq', 'model.layers.20.self_attn.rotary_emb.inv_freq', 'model.layers.1.self_attn.rotary_emb.inv_freq', 'model.layers.22.self_attn.rotary_emb.inv_freq', 'model.layers.0.self_attn.rotary_emb.inv_freq', 'model.layers.7.self_attn.rotary_emb.inv_freq', 'model.layers.15.self_attn.rotary_emb.inv_freq', 'model.layers.18.self_attn.rotary_emb.inv_freq', 'model.layers.2.self_attn.rotary_emb.inv_freq', 'model.layers.16.self_attn.rotary_emb.inv_freq', 'model.layers.14.self_attn.rotary_emb.inv_freq', 'model.lay

(…)2023/resolve/main/generation_config.json:   0%|          | 0.00/188 [00:00<?, ?B/s]

# Print model trainable parameters

In [7]:
def print_number_of_trainable_model_parameters(model):
    trainable_model_params = 0
    all_model_params = 0
    for _, param in model.named_parameters():
        all_model_params += param.numel()
        if param.requires_grad:
            trainable_model_params += param.numel()
    return f"trainable model parameters: {trainable_model_params}\nall model parameters: {all_model_params}\npercentage of trainable model parameters: {100 * trainable_model_params / all_model_params:.2f}%"

# print(print_number_of_trainable_model_parameters(original_model))

# Freeze all model parameters

In [8]:
# Freeze all parameters
for param in original_model.parameters():
    param.requires_grad = False

In [9]:
print(print_number_of_trainable_model_parameters(original_model))

trainable model parameters: 0
all model parameters: 6738415616
percentage of trainable model parameters: 0.00%



# Define prompt template

In [10]:
# define prompt format
def create_prompt(input):
    input_prompt = 'Input:'
    output_prompt = ', Output:'
    instruction = f"""Instruction: Given a search query, then route to different backend components based on the search intent.
1. If the search is about unit conversion, return API function UnitConvert(SourceUnit, TargetUnit, SourceValue).
2. If the search is about calculation, return API function Calculate(Equation).
3. If the search is about other search intent, return API function Search().
* For unit conversion: common unit conversion in length, mass, time, area, speed, temperature, volume should be covered. And it should be consistent for the same unit throughout. E.g. it should always be “foot”, it cannot be “feet” or “ft” in API calls.
* For calculation: common operation such as +, -, *, /, pow, log, ln, exp, tan(h), sin(h), cos(h), factorial should be covered. And it should be consistent for the same operation throughout. E.g. it should always be “ * ”, it cannot be “x” or “X” in API calls.
Handle input queries in different language styles. Cover common unit conversion and calculation operations.

"""

    prompt = instruction + input_prompt + f'“{input}”' + output_prompt
    return prompt

# Examples:
# {input_prompt}“ft to cm”{output_prompt}“UnitConvert(SourceUnit:foot, TargetUnit:centimeter,
# SourceValue:1)”
# {input_prompt}“how many ounces in 5.8 kilograms”{output_prompt}“UnitConvert(SourceUnit:kilogram,
# TargetUnit:ounce, SourceValue:5.8)”
# {input_prompt}“two to the power of 10”{output_prompt}“Calculate(2^10)”
# {input_prompt}“2001-1989” {output_prompt}“Calculate(2001-1989)”
# {input_prompt}“what is chatgpt”{output_prompt}“Search()”
# {input_prompt}“primary year 1 maths calculation checklist”{output_prompt}“Search()”
# {input_prompt}“what are different length units”{output_prompt}“Search()”
# {input_prompt}“Natural logarithm of -3/18”{output_prompt}“Calculate(ln(-3/18))”
# {input_prompt}“what is tan of 3/4”{output_prompt}“Calculate(tan(3/4))”


def create_prompt_training(input,output):
    input_prompt = 'Input:'
    output_prompt = ', Output:'
    start_prompt = '<s>[INST] '
    end_prompt = '[/INST]'
    instruction = f"""Instruction: Given a search query, then route to different backend components based on the search intent.
1. If the search is about unit conversion, return API function UnitConvert(SourceUnit, TargetUnit, SourceValue).
2. If the search is about calculation, return API function Calculate(Equation).
3. If the search is about other search intent, return API function Search().
* For unit conversion: common unit conversion in length, mass, time, area, speed, temperature, volume should be covered. And it should be consistent for the same unit throughout. E.g. it should always be “foot”, it cannot be “feet” or “ft” in API calls.
* For calculation: common operation such as +, -, *, /, pow, log, ln, exp, tan(h), sin(h), cos(h), factorial should be covered. And it should be consistent for the same operation throughout. E.g. it should always be “ * ”, it cannot be “x” or “X” in API calls.
Handle input queries in different language styles. Cover common unit conversion and calculation operations.

"""

    prompt = start_prompt + instruction + input_prompt + f'“{input}”' + output_prompt + f'“{output}”' + end_prompt
    return prompt

# # define prompt format
# def create_prompt(input):
#     start_prompt = 'Input:'#
#     end_prompt = ', Output:'#
#     instruction = f"""Instruction: Given a search query, then route to different backend components based on the search intent.
# 1. If the search is about unit conversion, return API function UnitConvert(SourceUnit, TargetUnit, SourceValue).
# 2. If the search is about calculation, return API function Calculate(Equation).
# 3. If the search is about other search intent, return API function Search().
# * For unit conversion: common unit conversion in length, mass, time, area, speed, temperature, volume should be covered. And it should be consistent for the same unit throughout. E.g. it should always be “foot”, it cannot be “feet” or “ft” in API calls.
# * For calculation: common operation such as +, -, *, /, pow, log, ln, exp, tan(h), sin(h), cos(h), factorial should be covered. And it should be consistent for the same operation throughout. E.g. it should always be “ * ”, it cannot be “x” or “X” in API calls.
# Handle input queries in different language styles. Cover common unit conversion and calculation operations.

# Examples:
# {start_prompt}“ft to cm”{end_prompt}“UnitConvert(SourceUnit:foot, TargetUnit:centimeter,
# SourceValue:1)”
# {start_prompt}“how many ounces in 5.8 kilograms”{end_prompt}“UnitConvert(SourceUnit:kilogram,
# TargetUnit:ounce, SourceValue:5.8)”
# {start_prompt}“two to the power of 10”{end_prompt}“Calculate(2^10)”
# {start_prompt}“2001-1989” {end_prompt}“Calculate(2001-1989)”
# {start_prompt}“what is chatgpt”{end_prompt}“Search()”
# {start_prompt}“primary year 1 maths calculation checklist”{end_prompt}“Search()”
# {start_prompt}“what are different length units”{end_prompt}“Search()”
# {start_prompt}“Natural logarithm of -3/18”{end_prompt}“Calculate(ln(-3/18))”
# {start_prompt}“what is tan of 3/4”{end_prompt}“Calculate(tan(3/4))”

# Only return output of the the given input.

# """
#     prompt = instruction + start_prompt + f'“{input}”' + end_prompt
#     return prompt

# Generate response in tokens from loaded model
The input_ids and generations are tokens

In [11]:
def model_generate(model, input):
    generation_config = model.generation_config
    generation_config.max_new_tokens = 100
    generation_config.temperature = 0.00000000000001
    generation_config.top_p = 0.9
    generation_config.num_return_sequences = 1
    generation_config.pad_token_id = tokenizer.eos_token_id
    generation_config.eos_token_id = tokenizer.eos_token_id
    return model.generate(
            input_ids = input.input_ids,
            attention_mask = input.attention_mask,
            generation_config = generation_config,
            )

# Generate respsons in text using giving input query text

In [12]:
def generated_text(model,input):
    prompt = create_prompt(input)
    inputs = tokenizer(prompt, return_tensors='pt')
    generated = tokenizer.decode(
        model_generate(model, inputs)[0],
        skip_special_tokens=True
    )
    return generated

# def generated_text(model,input):
#     input_prompt = 'Input:'
#     output_prompt = ', Output:'
#     prompt=input_prompt+input+output_prompt
#     inputs = tokenizer(prompt, return_tensors='pt')
#     generated = tokenizer.decode(
#         model_generate(model, inputs)[0],
#         skip_special_tokens=True
#     )
#     return generated

# Post processing generation output from LLM

In [40]:
def extractOutputString(input_string,output_string):
    import re
    # Use regular expressions to find the matching output for the input query
    output_match = re.search(rf'Input:\s*“{re.escape(input_string)}”\s*,\s*Output:\s*(.*)(?:[\[]|$)', output_string, flags=re.MULTILINE)
    print(output_match)
    # Extract and print the output
    if output_match:
        output_string = output_match.group(1)
    # else:
    #     output_match = re.search(rf'Input:\s*“{re.escape(input_string)}”\s*,\s*Output:\s*(.*)[\[]', output_string, flags=re.MULTILINE)
    #     output_string = output_match.group(1)

    # Remove quotation marks
    prefixes = ['“', '”', "'", '"', '[', '.', ']']
    if output_string.startswith(tuple(prefixes)):
        output_string = output_string[1:]
    while output_string.endswith(tuple(prefixes)):
        output_string = output_string[:-1]
    # Remove all space in output
    output_string = "".join(output_string.split())
    return output_string

# Try with original model

In [41]:
example_indices = [5, 9, 24, 78]#[49]#[78]#

dash_line = '-'.join('' for x in range(100))

for i, index in enumerate(example_indices):
    input = q2f_datasets['test'][index]['input']
    output = q2f_datasets['test'][index]['output']

    generated = generated_text(original_model,input)
    print(generated)
    generated = extractOutputString(input,generated)

    print(dash_line)
    print('Example ', i + 1)
    print(dash_line)
    print(f'INPUT:\n{input}')
    print(dash_line)
    print(f'BASELINE OUTPUT:\n{output}')
    print(dash_line)
    print(f'MODEL GENERATION - OUTPUT:\n{generated}\n')

Instruction: Given a search query, then route to different backend components based on the search intent.
1. If the search is about unit conversion, return API function UnitConvert(SourceUnit, TargetUnit, SourceValue).
2. If the search is about calculation, return API function Calculate(Equation).
3. If the search is about other search intent, return API function Search().
* For unit conversion: common unit conversion in length, mass, time, area, speed, temperature, volume should be covered. And it should be consistent for the same unit throughout. E.g. it should always be “foot”, it cannot be “feet” or “ft” in API calls.
* For calculation: common operation such as +, -, *, /, pow, log, ln, exp, tan(h), sin(h), cos(h), factorial should be covered. And it should be consistent for the same operation throughout. E.g. it should always be “ * ”, it cannot be “x” or “X” in API calls.
Handle input queries in different language styles. Cover common unit conversion and calculation operations.



# Pre-processing dataset for training
Tokenize inputs and outputs

In [42]:
from datasets import DatasetDict, Dataset

def create_text_datasets(example):
    # Define your custom processing logic here
    prompt_text = create_prompt_training(example['input'], example['output'])
    return {
        "text": prompt_text
    }

def process_dataset_dict(dataset_dict, processing_function):
    processed_dict = DatasetDict()

    for split_key, split_data in dataset_dict.items():
        processed_data = split_data.map(processing_function)
        processed_dict[split_key] = processed_data

    return processed_dict

# Concatenate "input" and "output" using the custom function
datasets = process_dataset_dict(q2f_datasets, create_text_datasets)
datasets

Map:   0%|          | 0/210 [00:00<?, ? examples/s]

Map:   0%|          | 0/50 [00:00<?, ? examples/s]

Map:   0%|          | 0/80 [00:00<?, ? examples/s]

DatasetDict({
    train: Dataset({
        features: ['input', 'output', 'text'],
        num_rows: 210
    })
    validation: Dataset({
        features: ['input', 'output', 'text'],
        num_rows: 50
    })
    test: Dataset({
        features: ['input', 'output', 'text'],
        num_rows: 80
    })
})

In [43]:
# Iterate over your training set and calculate the length of each sequence.
# For example, you can use the following code:
def get_max_seq_length(datasets):
    max_seq_length = 0
    for sequence in datasets['train']:
        text_len = len(sequence['text'])
        if text_len > max_seq_length:
            max_seq_length = text_len
    return max_seq_length+100

max_seq_length = get_max_seq_length(datasets)
max_seq_length

1276

In [45]:
# PEFT Setup
lora_config = LoraConfig(
    r=16,
    lora_alpha=32,
    target_modules=["q_proj", "k_proj", "v_proj", "o_proj"],#, "ffn_kernel"],
    lora_dropout=0.05,
    bias="none",
    task_type="CAUSAL_LM"
)
peft_model = original_model
# peft_model = get_peft_model(original_model, lora_config)
# print(print_number_of_trainable_model_parameters(peft_model))

output_dir = f'./peft-query-function-training-{str(int(time.time()))}'

peft_training_args = TrainingArguments(
    output_dir=output_dir,
    auto_find_batch_size=True,
    learning_rate=1e-3, # Higher learning rate than full fine-tuning.
    num_train_epochs=5,
    logging_steps=10,
    # max_steps=40
)

from trl import SFTTrainer
# Set supervised fine-tuning parameters
peft_trainer = SFTTrainer(
    model=peft_model,
    train_dataset=datasets['train'],
    eval_dataset=datasets['validation'],
    peft_config=lora_config,
    dataset_text_field="text",
    max_seq_length=max_seq_length,
    tokenizer=tokenizer,
    args=peft_training_args,
    packing=False,
)

peft_trainer.train()

peft_model_path="./peft-query-function-checkpoint-local"

peft_trainer.model.save_pretrained(peft_model_path,
                                   local_files_only=True
                                   )
tokenizer.save_pretrained(peft_model_path,
                          local_files_only=True
                          )

Map:   0%|          | 0/210 [00:00<?, ? examples/s]

Map:   0%|          | 0/50 [00:00<?, ? examples/s]

/usr/local/lib/python3.10/dist-packages/transformers/optimization.py:391: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  warnings.warn(
You're using a LlamaTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.
`use_cache=True` is incompatible with gradient checkpointing. Setting `use_cache=False`...
/usr/local/lib/python3.10/dist-packages/bitsandbytes/autograd/_functions.py:322: UserWarning: MatMul8bitLt: inputs will be cast from torch.float32 to float16 during quantization
  warnings.warn(f"MatMul8bitLt: inputs will be cast from {A.dtype} to float16 during quantization")


('./peft-query-function-checkpoint-local/tokenizer_config.json',
 './peft-query-function-checkpoint-local/special_tokens_map.json',
 './peft-query-function-checkpoint-local/tokenizer.json')

## Dataset Tokenisation

In [ ]:
model_max_length=150#394
def tokenize_function(example):
    prompt = [create_prompt(input) for input in example["input"]]
    example['input_ids'] = tokenizer(prompt, padding='max_length', truncation=True, max_length=model_max_length, return_tensors="pt").input_ids
    example['labels'] = tokenizer(example["output"], padding='max_length', truncation=True, max_length=model_max_length, return_tensors="pt").input_ids

    return example

# The dataset actually contains 3 diff splits: train, validation, test.
# The tokenize_function code is handling all data across all splits in batches.
tokenized_datasets = q2f_datasets.shuffle().map(tokenize_function, batched=True,batch_size=30)
tokenized_datasets = tokenized_datasets.remove_columns(['input', 'output'])
# tokenized_datasets = tokenized_datasets.remove_columns(['Search', 'input', 'output'])
tokenized_datasets

Map:   0%|          | 0/210 [00:00<?, ? examples/s]

Map:   0%|          | 0/50 [00:00<?, ? examples/s]

Map:   0%|          | 0/80 [00:00<?, ? examples/s]

DatasetDict({
    train: Dataset({
        features: ['input_ids', 'labels'],
        num_rows: 210
    })
    validation: Dataset({
        features: ['input_ids', 'labels'],
        num_rows: 50
    })
    test: Dataset({
        features: ['input_ids', 'labels'],
        num_rows: 80
    })
})

## PEFT training

In [ ]:
# PEFT Setup
lora_config = LoraConfig(
    r=16,
    lora_alpha=32,
    target_modules=["q_proj", "k_proj", "v_proj", "o_proj"],#, "ffn_kernel"],
    lora_dropout=0.05,
    bias="none",
    task_type="CAUSAL_LM"
)
peft_model = get_peft_model(original_model, lora_config)
print(print_number_of_trainable_model_parameters(peft_model))

trainable model parameters: 16777216
all model parameters: 6755192832
percentage of trainable model parameters: 0.25%


In [ ]:
output_dir = f'./peft-query-function-training-{str(int(time.time()))}'

peft_training_args = TrainingArguments(
    output_dir=output_dir,
    auto_find_batch_size=True,
    learning_rate=1e-2, # Higher learning rate than full fine-tuning.
    num_train_epochs=5,
    logging_steps=50,
    # max_steps=40
)

peft_trainer = Trainer(
    model=peft_model,
    args=peft_training_args,
    train_dataset=tokenized_datasets["train"],
    eval_dataset=tokenized_datasets['validation']
)

In [ ]:
peft_trainer.train()

peft_model_path="./peft-query-function-checkpoint-local"

peft_trainer.model.save_pretrained(peft_model_path)
tokenizer.save_pretrained(peft_model_path)

/usr/local/lib/python3.10/dist-packages/transformers/optimization.py:391: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  warnings.warn(


('./peft-query-function-checkpoint-local/tokenizer_config.json',
 './peft-query-function-checkpoint-local/special_tokens_map.json',
 './peft-query-function-checkpoint-local/tokenizer.json')

In [ ]:
!ls -al ./peft-query-function-checkpoint-local/adapter_model.bin

-rw-r--r-- 1 root root 1048576 Oct 14 06:23 ./peft-query-function-checkpoint-local/adapter_model.bin


In [ ]:
del original_model

NameError: ignored

In [ ]:
# peft_model_base = load_model(model_name)
peft_model_base = AutoModelForCausalLM.from_pretrained(model_name)
# peft_model_base=original_model
tokenizer = AutoTokenizer.from_pretrained(model_name)#, torch_dtype=torch.bfloat16)
peft_model_path="./peft-query-function-checkpoint-local"
peft_model = PeftModel.from_pretrained(peft_model_base,
                                       peft_model_path,
                                      #  torch_dtype=torch.bfloat16,
                                       is_trainable=False)
tokenizer = AutoTokenizer.from_pretrained(peft_model_path)
print(print_number_of_trainable_model_parameters(peft_model))

ValueError: ignored

## Finetune without lora

In [ ]:
output_dir = f'./finetune-query-function-training-{str(int(time.time()))}'

finetune_model=original_model

model_training_args = TrainingArguments(
    output_dir=output_dir,
    auto_find_batch_size=True,
    learning_rate=1e-3,
    num_train_epochs=1,
    logging_steps=10,
    max_steps=40,
    # optim="adamw_torch"
)

finetune_trainer = Trainer(
    model=finetune_model,
    args=model_training_args,
    train_dataset=tokenized_datasets["train"],
    eval_dataset=tokenized_datasets['validation']
)

In [ ]:
# Free GPU memory
torch.cuda.empty_cache()

finetune_trainer.train()

finetune_model_path="./finetune-query-function-checkpoint-local"
finetune_trainer.model.save_pretrained(finetune_model_path,
                                       local_files_only=True
                                       )
tokenizer.save_pretrained(finetune_model_path,
                          local_files_only=True
                          )

OutOfMemoryError: ignored

### Check model state dict

In [ ]:
state_dict=finetune_model.state_dict()
for key, weight in state_dict.items():
    print(key)
    print(weight)
    # weight_size = weight.numel() * dtype_byte_size(weight.dtype)

finetune_model.state_dict()

model.embed_tokens.weight
tensor([[nan, nan, nan,  ..., nan, nan, nan],
        [nan, nan, nan,  ..., nan, nan, nan],
        [nan, nan, nan,  ..., nan, nan, nan],
        ...,
        [nan, nan, nan,  ..., nan, nan, nan],
        [nan, nan, nan,  ..., nan, nan, nan],
        [nan, nan, nan,  ..., nan, nan, nan]], device='cuda:0',
       dtype=torch.float16)
model.layers.0.self_attn.q_proj.weight
tensor([[ -6, -19,  -2,  ...,   6,   6,  14],
        [  4,   7,   2,  ..., -33, -28,  -5],
        [ -3,   1,  -1,  ...,  -6, -22,  15],
        ...,
        [ 19,  -2,   0,  ...,   2,  -9,   2],
        [ -3,   3,  -3,  ...,   3,   6,   0],
        [  2,  -2,   0,  ...,  15,  15,  -7]], device='cuda:0',
       dtype=torch.int8)
model.layers.0.self_attn.q_proj.SCB
tensor([0.1155, 0.2032, 0.2490,  ..., 0.2150, 0.1587, 0.1527], device='cuda:0')
model.layers.0.self_attn.q_proj.weight_format
col_turing
model.layers.0.self_attn.k_proj.weight
tensor([[ -7,   3,   0,  ...,   8,  -6,  -4],
        [ 

OrderedDict([('model.embed_tokens.weight',
              tensor([[nan, nan, nan,  ..., nan, nan, nan],
                      [nan, nan, nan,  ..., nan, nan, nan],
                      [nan, nan, nan,  ..., nan, nan, nan],
                      ...,
                      [nan, nan, nan,  ..., nan, nan, nan],
                      [nan, nan, nan,  ..., nan, nan, nan],
                      [nan, nan, nan,  ..., nan, nan, nan]], device='cuda:0',
                     dtype=torch.float16)),
             ('model.layers.0.self_attn.q_proj.weight',
              tensor([[ -6, -19,  -2,  ...,   6,   6,  14],
                      [  4,   7,   2,  ..., -33, -28,  -5],
                      [ -3,   1,  -1,  ...,  -6, -22,  15],
                      ...,
                      [ 19,  -2,   0,  ...,   2,  -9,   2],
                      [ -3,   3,  -3,  ...,   3,   6,   0],
                      [  2,  -2,   0,  ...,  15,  15,  -7]], device='cuda:0',
                     dtype=torch.int8)),
      

# Load finetuned model for inference and eveluate test dataset

In [ ]:
import os
os.kill(os.getpid(), 9)

In [ ]:
import os
import time
import pandas as pd
import torch
import evaluate
import tqdm

from transformers import AutoModelForCausalLM, AutoTokenizer, BitsAndBytesConfig, GenerationConfig, TrainingArguments

from peft import PeftModel, LoraConfig, get_peft_model, TaskType

In [ ]:
tokenizer = AutoTokenizer.from_pretrained(finetune_model_path, torch_dtype=torch.bfloat16)
instruct_model = AutoModelForCausalLM.from_pretrained(finetune_model_path, torch_dtype=torch.bfloat16)

OSError: ignored

# PEFT model load for inference and eveluate test dataset
Require restart runtime due to memory issue

In [ ]:
import os
os.kill(os.getpid(), 9)

In [1]:
import os
import time
import pandas as pd
import torch
import evaluate
import tqdm

from transformers import AutoModelForCausalLM, AutoTokenizer, BitsAndBytesConfig, GenerationConfig, TrainingArguments

from peft import PeftModel, LoraConfig, get_peft_model, TaskType

In [3]:
### Load datasets
from SearchQuery2FuncCall.setup_dataset import text2json, load_n_process_data

text2json('/content/SearchQuery2FuncCall/Dataset.txt')
# q2f_datasets = load_n_process_data('/content/non_search_examples.json')
q2f_datasets = load_n_process_data('/content/q2f_dataset.json')
q2f_datasets


### Access Huggingface for loading model
""" # Get model api key """
def load_api_key_from_file(file_path):
    with open(file_path, 'r') as file:
        api_key = file.read().strip()
    return api_key

# Setting a new environment variable
os.environ["HUGGINGFACE_TOKEN"] = load_api_key_from_file('/content/api_key_huggingface.txt')

!huggingface-cli login --token $HUGGINGFACE_TOKEN


### Funciton to print number of trainable mode parameters
def print_number_of_trainable_model_parameters(model):
    trainable_model_params = 0
    all_model_params = 0
    for _, param in model.named_parameters():
        all_model_params += param.numel()
        if param.requires_grad:
            trainable_model_params += param.numel()
    return f"trainable model parameters: {trainable_model_params}\nall model parameters: {all_model_params}\npercentage of trainable model parameters: {100 * trainable_model_params / all_model_params:.2f}%"


Saved 340 examples to 'q2f_dataset.json'.
Separated 87 Search() examples to 'search_examples.json'.
Separated 253 non-Search() examples to 'non_search_examples.json'.


Extracting data files:   0%|          | 0/1 [00:00<?, ?it/s]

Generating train split: 0 examples [00:00, ? examples/s]

Dataset json downloaded and prepared to /root/.cache/huggingface/datasets/json/default-7148424e417eeb6b/0.0.0/fe5dd6ea2639a6df622901539cb550cf8797e5a6b2dd7af1cf934bed8e233e6e. Subsequent calls will reuse this data.


  0%|          | 0/1 [00:00<?, ?it/s]

Token will not been saved to git credential helper. Pass `add_to_git_credential=True` if you want to set the git credential as well.
Token is valid (permission: read).
Your token has been saved to /root/.cache/huggingface/token
Login successful


In [2]:
### Load PEFT model from previous saved folder
# Note: It will only work if the colab session is not disconnected from previous training session
# Restart the runtime is fine
model_name = "atwine/llama-2-7b-chat-fully-quantized-q4-06092023"
peft_model_base = AutoModelForCausalLM.from_pretrained(model_name, torch_dtype=torch.bfloat16)
peft_model_path="./peft-query-function-checkpoint-local"
peft_model = PeftModel.from_pretrained(peft_model_base,
                                       peft_model_path,
                                       torch_dtype=torch.bfloat16,
                                       is_trainable=False
                                      )
# peft_model = peft_model.merge_and_unload()
tokenizer = AutoTokenizer.from_pretrained(peft_model_path, torch_dtype=torch.bfloat16)
print(print_number_of_trainable_model_parameters(peft_model))

Some weights of LlamaForCausalLM were not initialized from the model checkpoint at atwine/llama-2-7b-chat-fully-quantized-q4-06092023 and are newly initialized: ['model.layers.29.self_attn.rotary_emb.inv_freq', 'model.layers.20.self_attn.rotary_emb.inv_freq', 'model.layers.2.self_attn.rotary_emb.inv_freq', 'model.layers.6.self_attn.rotary_emb.inv_freq', 'model.layers.22.self_attn.rotary_emb.inv_freq', 'model.layers.30.self_attn.rotary_emb.inv_freq', 'model.layers.18.self_attn.rotary_emb.inv_freq', 'model.layers.12.self_attn.rotary_emb.inv_freq', 'model.layers.14.self_attn.rotary_emb.inv_freq', 'model.layers.26.self_attn.rotary_emb.inv_freq', 'model.layers.16.self_attn.rotary_emb.inv_freq', 'model.layers.10.self_attn.rotary_emb.inv_freq', 'model.layers.3.self_attn.rotary_emb.inv_freq', 'model.layers.17.self_attn.rotary_emb.inv_freq', 'model.layers.1.self_attn.rotary_emb.inv_freq', 'model.layers.5.self_attn.rotary_emb.inv_freq', 'model.layers.25.self_attn.rotary_emb.inv_freq', 'model.lay

NameError: ignored

In [53]:
### Prepare for model generation
# define prompt format
def create_prompt(input):
    input_prompt = 'Input:'
    output_prompt = ', Output:'
    instruction = f"""Instruction: Given a search query, then route to different backend components based on the search intent.
1. If the search is about unit conversion, return API function UnitConvert(SourceUnit, TargetUnit, SourceValue).
2. If the search is about calculation, return API function Calculate(Equation).
3. If the search is about other search intent, return API function Search().
* For unit conversion: common unit conversion in length, mass, time, area, speed, temperature, volume should be covered. And it should be consistent for the same unit throughout. E.g. it should always be “foot”, it cannot be “feet” or “ft” in API calls.
* For calculation: common operation such as +, -, *, /, pow, log, ln, exp, tan(h), sin(h), cos(h), factorial should be covered. And it should be consistent for the same operation throughout. E.g. it should always be “ * ”, it cannot be “x” or “X” in API calls.
Handle input queries in different language styles. Cover common unit conversion and calculation operations.

"""
    prompt = instruction + input_prompt + f'“{input}”' + output_prompt
    return prompt

# Generate response in tokens with given model and tokenized input
def model_generate(model, input):
    generation_config = model.generation_config
    generation_config.max_new_tokens = 100
    generation_config.temperature = 0.00000000000001
    generation_config.top_p = 0.9
    generation_config.num_return_sequences = 1
    generation_config.pad_token_id = tokenizer.eos_token_id
    generation_config.eos_token_id = tokenizer.eos_token_id
    return model.generate(
            input_ids = input.input_ids,
            attention_mask = input.attention_mask,
            generation_config = generation_config,
            )

# Generate Text with selected model and input text
def generated_text(model,input):
    prompt = create_prompt(input)
    inputs = tokenizer(prompt, return_tensors='pt')
    generated = tokenizer.decode(
        model_generate(model, inputs)[0],
        skip_special_tokens=True
    )
    return generated

# Post generation string processing
def extractOutputString(input_string,output_string):
    import re
    # Use regular expressions to find the matching output for the input query
    # output_match = re.search(rf'Input:\s*“{re.escape(input_string)}”\s*,\s*Output:\s*(.*)\[\/', output_string, flags=re.MULTILINE)
    output_match = re.search(rf'Input:\s*“{re.escape(input_string)}”\s*,\s*Output:\s*(.*?)(\[\/|$)', output_string, flags=re.MULTILINE)

    # Extract and print the output
    if output_match:
        output_string = output_match.group(1)
    # else:
    #     output_match = re.search(rf'Input:\s*“{re.escape(input_string)}”\s*,\s*Output:\s*(.*)$', output_string, flags=re.MULTILINE)
    #     if output_match:
    #         output_string = output_match.group(1)

    # Remove quotation marks
    prefixes = ['“', '”', "'", '"', '[', '.', ']']
    if output_string.startswith(tuple(prefixes)):
        output_string = output_string[1:]
    while output_string.endswith(tuple(prefixes)):
        output_string = output_string[:-1]
    # Remove all space in output
    output_string = "".join(output_string.split())
    return output_string

# Generate with selected examples

In [55]:
example_indices = [5, 9, 24, 78]#[49]#

dash_line = '-'.join('' for x in range(100))

for i, index in enumerate(example_indices):
    input = q2f_datasets['test'][index]['input']
    output = q2f_datasets['test'][index]['output']

    generated_peft = generated_text(peft_model,input)
    generated_peft = extractOutputString(input,generated_peft)

    print(dash_line)
    print('Example ', i + 1)
    print(dash_line)
    print(f'INPUT:\n{input}')
    print(dash_line)
    print(f'BASELINE OUTPUT:\n{output}')
    print(dash_line)
    print(f'PEFT MODEL GENERATION - OUTPUT:\n{generated_peft}\n')

/usr/local/lib/python3.10/dist-packages/transformers/generation/utils.py:1405: UserWarning: You are calling .generate() with the `input_ids` being on a device type different than your model's device. `input_ids` is on cpu, whereas the model is on cuda. You may experience unexpected behaviors or slower generation. Please make sure that you have put `input_ids` to the correct device by calling for example input_ids = input_ids.to('cuda') before running `.generate()`.
  warnings.warn(


Instruction: Given a search query, then route to different backend components based on the search intent.
1. If the search is about unit conversion, return API function UnitConvert(SourceUnit, TargetUnit, SourceValue).
2. If the search is about calculation, return API function Calculate(Equation).
3. If the search is about other search intent, return API function Search().
* For unit conversion: common unit conversion in length, mass, time, area, speed, temperature, volume should be covered. And it should be consistent for the same unit throughout. E.g. it should always be “foot”, it cannot be “feet” or “ft” in API calls.
* For calculation: common operation such as +, -, *, /, pow, log, ln, exp, tan(h), sin(h), cos(h), factorial should be covered. And it should be consistent for the same operation throughout. E.g. it should always be “ * ”, it cannot be “x” or “X” in API calls.
Handle input queries in different language styles. Cover common unit conversion and calculation operations.



# Generate outputs for given test dataset
Here we are using the test dataset

In [61]:
index_s=0
index_e=index_s+len(q2f_datasets['test'])
inputs = q2f_datasets['test'][index_s:index_e]['input']
outputs = q2f_datasets['test'][index_s:index_e]['output']
# inputs = q2f_datasets['test']['input']
# outputs = q2f_datasets['test']['output']

API_outputs = []
API_outputs_peft = []

for idx, input in enumerate(inputs):
    if idx%10==0:
        print(idx)
    # API_output = generated_text(original_model,input)
    # API_output = extractOutputString(input,API_output)
    # API_outputs.append(API_output)
    API_output_peft = generated_text(peft_model,input)
    API_output_peft = extractOutputString(input,API_output_peft)
    API_outputs_peft.append(API_output_peft)


# zipped_summaries = list(zip(inputs, outputs, API_outputs, API_outputs_peft))
# df = pd.DataFrame(zipped_summaries, columns = ['inputs', 'outputs', 'API_outputs', 'API_outputs_peft'])

zipped_summaries = list(zip(inputs, outputs, API_outputs_peft))
df = pd.DataFrame(zipped_summaries, columns = ['inputs', 'outputs', 'API_outputs_peft'])
df

0
<re.Match object; span=(1000, 1048), match='Input:“how to bake a pizza”, Output:“Search()”[/'>
<re.Match object; span=(1000, 1052), match='Input:“Best movies of all time”, Output:“Search()>
<re.Match object; span=(1000, 1061), match='Input:“natural logarithm of 10”, Output:“Calculat>
<re.Match object; span=(1000, 1055), match='Input:“top 10 travel destinations”, Output:“Searc>
<re.Match object; span=(1000, 1118), match='Input:“Convert 2 acres to square meters”, Output:>
<re.Match object; span=(1000, 1058), match='Input:“logarithm of 1000”, Output:“Calculate(log(>
<re.Match object; span=(1000, 1058), match='Input:“interesting facts about space”, Output:“Se>
<re.Match object; span=(1000, 1116), match='Input:“cubic feet to cubic meters”, Output:“UnitC>
<re.Match object; span=(1000, 1059), match='Input:“cosine of 45 degrees”, Output:“Calculate(c>
<re.Match object; span=(1000, 1056), match='Input:“inventions by Thomas Edison”, Output:“Sear>
10
<re.Match object; span=(1000, 1071), match='I

inputs  \
0                how to bake a pizza   
1            Best movies of all time   
2            natural logarithm of 10   
3         top 10 travel destinations   
4   Convert 2 acres to square meters   
..                               ...   
75    Evaluate the expression 10 + 3   
76         Recipe for chocolate cake   
77                square root of 100   
78                   15 divided by 3   
79     compute the square root of 25   

                                              outputs  \
0                                            Search()   
1                                            Search()   
2                                   Calculate(ln(10))   
3                                            Search()   
4   UnitConvert(SourceUnit:acre,TargetUnit:squarem...   
..                                                ...   
75                                    Calculate(10+3)   
76                                           Search()   
77                               Calculate(sqrt(100))   
78                                    Calculate(15/3)   
79                                Calculate(sqrt(25))   

                                     API_outputs_peft  
0                                            Search()  
1                                            Search()  
2                                   Calculate(ln(10))  
3                                            Search()  
4   UnitConvert(SourceUnit:acre,TargetUnit:squarem...  
..                                                ...  
75                                    Calculate(10+3)  
76                                           Search()  
77                               Calculate(sqrt(100))  
78                                    Calculate(15/3)  
79                                Calculate(sqrt(25))  

[80 rows x 3 columns]

# Evaluate using ROUGE and BLEU scores

In [64]:
API_outputs=API_outputs_peft
# Rouge
rouge = evaluate.load('rouge')
API_model_results = rouge.compute(
    predictions=API_outputs,
    references=outputs[0:len(API_outputs)],
    use_aggregator=True,
    use_stemmer=True,
)

print('API MODEL ROUGE SCORES:')
print(API_model_results)

# bleu
bleu = evaluate.load('bleu')
API_model_results = bleu.compute(
    predictions=API_outputs,
    references=outputs,
)

print('API MODEL BLEU SCORES:')
print(API_model_results)

API MODEL ROUGE SCORES:
{'rouge1': 0.991875, 'rouge2': 0.7375, 'rougeL': 0.991875, 'rougeLsum': 0.991875}
API MODEL BLEU SCORES:
{'bleu': 0.9727731466373349, 'precisions': [0.9926793557833089, 0.9867330016583747, 0.9789674952198852, 0.9729119638826185], 'brevity_penalty': 0.9898034391264364, 'length_ratio': 0.9898550724637681, 'translation_length': 683, 'reference_length': 690}


## Evaluation Results

MODEL ROUGE SCORES:
{'rouge1': 0.991875, 'rouge2': 0.7375, 'rougeL': 0.991875, 'rougeLsum': 0.991875}

MODEL BLEU SCORES:
{'bleu': 0.9727731466373349, 'precisions': [0.9926793557833089, 0.9867330016583747, 0.9789674952198852, 0.9729119638826185], 'brevity_penalty': 0.9898034391264364, 'length_ratio': 0.9898550724637681, 'translation_length': 683, 'reference_length': 690}